cmd/thanos/receive.go

```diff
@@ -193,29 +193,28 @@ func runReceive(
 	{
 		// TSDB.
 		cancel := make(chan struct{})
-		// Before actually starting, we need to make sure
-		// the WAL is flushed.
-		startTimeMargin := int64(2 * time.Duration(tsdbCfg.MinBlockDuration).Seconds() * 1000)
-		if err := db.Open(); err != nil {
-			return errors.Wrap(err, "opening storage")
-		}
-		if err := db.Flush(); err != nil {
-			return errors.Wrap(err, "flushing storage")
-		}
 		g.Add(func() error {
 			defer close(dbReady)
 			defer close(uploadC)
 
+			// Before actually starting, we need to make sure the
+			// WAL is flushed. The WAL is flushed after the
+			// hashring ring is loaded.
+			startTimeMargin := int64(2 * time.Duration(tsdbCfg.MinBlockDuration).Seconds() * 1000)
+			if err := db.Open(); err != nil {
+				return errors.Wrap(err, "opening storage")
+			}
+
 			// Before quitting, ensure the WAL is flushed and the DB is closed.
 			defer func() {
-				if err := db.Flush(); err != nil {
-					level.Warn(logger).Log("err", err, "msg", "failed to flush storage")
-					return
-				}
 				if err := db.Close(); err != nil {
 					level.Warn(logger).Log("err", err, "msg", "failed to close storage")
 					return
 				}
+				if err := db.Flush(); err != nil {
```

Comment:
Even if the current implementation allows that a flush can be called after `db.Close` I don't think any part of the code should use it this way. I don't think that should use any api after calling `db.Close`

Reply:
Sure.. A question then, replaying the WAL on each open is quite expensive for us since we are ingesting hundreds of thousands of samples per send (potentially more in the the future).

Is there any impact on chunk upload if were to reduce the `tsdb.block-duration` by an order of magnitude. Say to 30m or smaller? This should help with overall memory consumption and WAL replay speeds.

Reply:
The WAL is completely empty after every flush (in fact the directory doesn’t exist). So there shouldn’t be any WAL-replaying cost when re-opening the DB after a flush.